### PR TITLE
IGVF-1934-add-funtional-assay-mechanism

### PR DESCRIPTION
**Pull request by shengchd**

### CODE REVIEW

File: src/igvfd/schemas/measurement_set.json [hunk 1]
```diff
@@ -269,6 +269,22 @@
                 "appscript": "[\"ENSMUSG00000044487\", \"john-doe:gene01\"]",
                 "igvf_utils": "ENSMUSG00000044487, john-doe:gene01"
             }
+        },
+        "functional_assay_mechanisms": {
+            "title": "Functional Assay Mechanisms",
+            "description": "The biological mechanism(s) this functional assay measures. For example, VAMP-seq (MultiSTEP) measures variant effects on protein carboxylation and secretion.",
```
Comment (meenakshikagda) [inline src/igvfd/schemas/measurement_set.json:275]: Maybe change mechanisms to processes? In GO terms we use the word Biological process so maybe it is more equivalent. "The biological processes measured by this functional assay. For example, a VAMP-seq (MultiSTEP) assay measures the effects of variants on protein carboxylation and secretion processes."

File: src/igvfd/schemas/analysis_set.json [hunk 2]
```diff
@@ -106,7 +106,8 @@
         "samples.sample_terms.term_name",
         "samples.targeted_sample_term.term_name",
         "submitted_by.title",
-        "summary"
+        "summary",
+        "functional_assay_mechanisms.term_name"
```
Comment (meenakshikagda) [inline src/igvfd/schemas/analysis_set.json:110]: Move it above samples.summary, so this list is in the alphabetical order.

File: src/igvfd/schemas/measurement_set.json [hunk 3]
```diff
@@ -269,6 +269,22 @@
                 "appscript": "[\"ENSMUSG00000044487\", \"john-doe:gene01\"]",
                 "igvf_utils": "ENSMUSG00000044487, john-doe:gene01"
             }
+        },
+        "functional_assay_mechanisms": {
```
Comment (meenakshikagda) [inline src/igvfd/schemas/measurement_set.json:273]: Did we ask the lab to come up with this name? or did  we coin it? Just curious if we could use some other term for mechanisms. Maybe it is okay

File: src/igvfd/tests/fixtures/schemas/measurement_set.py [hunk 4]
```diff
@@ -275,3 +275,17 @@ def measurement_set_with_protocols(testapp, lab, award, assay_term_starr, tissue
...
+    item = {
+        'award': award['@id'],
+        'lab': lab['@id'],
+        'assay_term': assay_term_starr['@id'],
+        'samples': [tissue['@id']],
+        'file_set_type': 'experimental data',
+        'preferred_assay_title': 'SUPERSTARR',
+        'functional_assay_mechanisms': [phenotype_term_from_go['@id']]
+    }
+    return testapp.post_json('/measurement_set', item).json['@graph'][0]
```
Comment (meenakshikagda) [inline src/igvfd/tests/fixtures/schemas/measurement_set.py:291]: Do we need to add status=201 here? Interestingly none of the measurement-set fixtures have this.

File: src/igvfd/tests/fixtures/schemas/measurement_set.py [hunk 5]
```diff
@@ -275,3 +275,17 @@ def measurement_set_with_protocols(testapp, lab, award, assay_term_starr, tissue
...
+
+@pytest.fixture
+def measurement_set_with_functional_assay_mechanisms(testapp, lab, award, assay_term_starr, tissue, phenotype_term_from_go):
+    item = {
+        'award': award['@id'],
+        'lab': lab['@id'],
+        'assay_term': assay_term_starr['@id'],
+        'samples': [tissue['@id']],
+        'file_set_type': 'experimental data',
+        'preferred_assay_title': 'SUPERSTARR',
```
Comment (meenakshikagda) [inline src/igvfd/tests/fixtures/schemas/measurement_set.py:288]: Let's change this to VAMP-seq (Multi step) to have a consistent example.

File: src/igvfd/tests/fixtures/schemas/measurement_set.py [hunk 6]
```diff
@@ -275,3 +275,17 @@ def measurement_set_with_protocols(testapp, lab, award, assay_term_starr, tissue
...
     }
     return testapp.post_json('/measurement_set', item).json['@graph'][0]
+
+
+@pytest.fixture
+def measurement_set_with_functional_assay_mechanisms(testapp, lab, award, assay_term_starr, tissue, phenotype_term_from_go):
+    item = {
+        'award': award['@id'],
+        'lab': lab['@id'],
+        'assay_term': assay_term_starr['@id'],
```
Comment (meenakshikagda) [inline src/igvfd/tests/fixtures/schemas/measurement_set.py:285]: same as preferred_assay title, if we have assay_term for vamp-seq let's use that? Not too critical but good to have. :)

File: src/igvfd/tests/test_types_analysis_set.py [hunk 7]
```diff
@@ -121,3 +121,15 @@ def test_protocols(testapp, analysis_set_base, measurement_set_with_protocols):
     )
     res = testapp.get(analysis_set_base['@id'])
     assert res.json.get('protocols') == ['https://www.protocols.io/test-protocols-url-12345']
+
+
+def test_functional_assay_mechanisms(testapp, analysis_set_base, measurement_set_with_functional_assay_mechanisms, phenotype_term_from_go):
+    testapp.patch_json(
+        analysis_set_base['@id'],
+        {
+            'input_file_sets': [measurement_set_with_functional_assay_mechanisms['@id']]
```
Comment (meenakshikagda) [inline src/igvfd/tests/test_types_analysis_set.py:130]: Would it be a good idea to add two different measurement-sets with the same go_term? This way the with the test we can ensure we get a unique list of phenotype-terms on analysis-set.

File: src/igvfd/types/file_set.py [hunk 8]
```diff
@@ -317,6 +318,30 @@ def protocols(self, request, input_file_sets=None):
                     protocols.update(protocol)
         return list(protocols)
 
+    @calculated_property(
+        schema={
+            'title': 'Functional Assay Mechanisms',
+            'description': 'The biological mechanism(s) the functional assay measures.',
```
Comment (meenakshikagda) [inline src/igvfd/types/file_set.py:324]: See comment in measurement-set. Sentence 1 can be copied from there.

File: src/igvfd/types/file_set.py [hunk 9]
```diff
@@ -317,6 +318,30 @@ def protocols(self, request, input_file_sets=None):
...
+            'items': {
+                'title': 'Phenotype Term',
+                'type': 'string',
+                'linkTo': 'PhenotypeTerm'
+            },
+            'notSubmittable': True
+        }
+    )
+    def functional_assay_mechanisms(self, request, input_file_sets=None):
+        '''Calculate an array of unique biological mechanism(s) measured by the funcional assays associated with an analysis set.'''
```
Comment (meenakshikagda) [inline src/igvfd/types/file_set.py:337]: I like to have comments in the code but for IGVFd and encodeD we have avoided them usually :)

**Review by meenakshikagda — COMMENTED** [review]
Added some comments.

File: src/igvfd/schemas/analysis_set.json (hunk 2 above)
Comment (shengchd) [reply src/igvfd/schemas/analysis_set.json:110]: will reorder it

File: src/igvfd/schemas/measurement_set.json (hunk 3 above)
Comment (shengchd) [reply src/igvfd/schemas/measurement_set.json:273]: ah yes, as I replied in the first comment, they suggested the name.

File: src/igvfd/tests/fixtures/schemas/measurement_set.py (hunk 5 above)
Comment (shengchd) [reply src/igvfd/tests/fixtures/schemas/measurement_set.py:288]: yeah I will need to add an assay term for VAMP-seq in fixtures then

File: src/igvfd/tests/fixtures/schemas/measurement_set.py (hunk 6 above)
Comment (shengchd) [reply src/igvfd/tests/fixtures/schemas/measurement_set.py:285]: yes same as replied in the previous comment

File: src/igvfd/types/file_set.py (hunk 9 above)
Comment (shengchd) [reply src/igvfd/types/file_set.py:337]: I will remove it

File: src/igvfd/schemas/measurement_set.json [hunk 10]
```diff
@@ -269,6 +269,22 @@
...
+            "title": "Functional Assay Mechanisms",
+            "description": "The biological process(es) measured by this functional assay. For example, a VAMP-seq (MultiSTEP) assay measures the effects of variants on protein carboxylation and secretion processes.",
```
Comment (meenakshikagda) [inline src/igvfd/schemas/measurement_set.json:275]: Change process(es) to just processes

File: src/igvfd/types/file_set.py [hunk 11]
```diff
@@ -433,6 +434,29 @@ def sample_summary(self, request, samples=None):
...
+            'items': {
+                'title': 'Phenotype Term',
+                'type': 'string',
+                'linkTo': 'PhenotypeTerm'
+            },
+            'notSubmittable': True
+        }
+    )
+    def functional_assay_mechanisms(self, request, input_file_sets=None):
+        mechanism_objs = []
```
Comment (meenakshikagda) [inline src/igvfd/types/file_set.py:453]: let's use objects instead of objs

File: src/igvfd/types/file_set.py [hunk 12]
```diff
@@ -433,6 +434,29 @@ def sample_summary(self, request, samples=None):
...
+        file_set_objs = get_fileset_objs_from_input_file_sets(request=request, input_file_sets=input_file_sets)
```
Comment (meenakshikagda) [inline src/igvfd/types/file_set.py:454]: objs is a bit unclear. Let's make it objects instead

**Pull request merged**